src/trustchain.rs

```diff
@@ -122,3 +126,239 @@ impl TrustchainContext {
+        }
+    }
+
+    #[getter]
+    fn user_id(&self) -> PyResult<UserID> {
+        match &self.0 {
+            libparsec::core::TrustchainError::InvalidSelfSignedUserCertificate { user_id } => {
+                Ok(ids::UserID(user_id.clone()))
+            }
+            libparsec::core::TrustchainError::InvalidSelfSignedUserRevocationCertificate {
+                user_id,
+            } => Ok(ids::UserID(user_id.clone())),
+            libparsec::core::TrustchainError::InvalidSignatureGiven { user_id, .. } => {
+                Ok(ids::UserID(user_id.clone()))
+            }
+            libparsec::core::TrustchainError::MissingUserCertificate { user_id, .. } => {
+                Ok(ids::UserID(user_id.clone()))
+            }
+            _ => Err(PyAttributeError::new_err("No such attribute `user_id`")),
+        }
+    }
+
+    #[getter]
+    fn device_id(&self) -> PyResult<ids::DeviceID> {
+        if let libparsec::core::TrustchainError::MissingDeviceCertificate { device_id, .. } =
+            &self.0
+        {
+            Ok(ids::DeviceID(device_id.clone()))
+        } else {
+            Err(PyAttributeError::new_err("No such attribute `device_id`"))
+        }
+    }
+
+    #[getter]
+    fn verified_timestamp(&self) -> PyResult<DateTime> {
+        if let libparsec::core::TrustchainError::SignaturePosteriorUserRevocation {
+            verified_timestamp,
+            ..
+        } = &self.0
+        {
+            Ok(time::DateTime(*verified_timestamp))
+        } else {
+            Err(PyAttributeError::new_err(
+                "No such attribute `verified_timestamp`",
+            ))
+        }
+    }
+
+    #[getter]
+    fn user_timestamp(&self) -> PyResult<DateTime> {
+        if let libparsec::core::TrustchainError::SignaturePosteriorUserRevocation {
+            user_timestamp,
+            ..
+        } = &self.0
+        {
+            Ok(time::DateTime(*user_timestamp))
+        } else {
+            Err(PyAttributeError::new_err(
+                "No such attribute `user_timestamp`",
+            ))
+        }
+    }
+
+    #[getter]
+    fn expected(&self) -> PyResult<ids::UserID> {
+        if let libparsec::core::TrustchainError::UnexpectedCertificate { expected, .. } = &self.0 {
+            Ok(ids::UserID(expected.clone()))
+        } else {
+            Err(PyAttributeError::new_err("No such attribute `expected`"))
+        }
+    }
+
+    #[getter]
+    fn got(&self) -> PyResult<ids::UserID> {
+        if let libparsec::core::TrustchainError::UnexpectedCertificate { got, .. } = &self.0 {
+            Ok(ids::UserID(got.clone()))
+        } else {
+            Err(PyAttributeError::new_err("No such attribute `got`"))
+        }
+    }
+}
+
+gen_proto!(TrustchainError, __richcmp__, eq);
+gen_proto!(TrustchainError, __str__); // Needed for python's exceptions
+gen_proto!(TrustchainError, __repr__);
+
+impl From<TrustchainError> for PyErr {
```

Comment:
Don't we want to use `TrustchainErrorException` here ?

Reply:
This impl isn't used at all. We should just remove it